File: aes/AES.java
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.util.Arrays;

import javax.crypto.Cipher;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.Key;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.Security;
import java.util.Base64;

/**
 * @author ngh
 * AES128 算法
 *
 * CBC 模式
 *
 * PKCS7Padding 填充模式
 *
 * CBC模式需要添加一个参数iv
 *
 * 介于java 不支持PKCS7Padding，只支持PKCS5Padding 但是PKCS7Padding 和 PKCS5Padding 没有什么区别
 * 要实现在java端用PKCS7Padding填充，需要用到bouncycastle组件来实现
 * http://www.bouncycastle.org/archive/139/bcprov-jdk16-139.jar
 */
public class AES {
    // 算法名称
    final String KEY_ALGORITHM = "AES";
    // 加解密算法/模式/填充方式
    final String algorithmStr = "AES/CBC/PKCS7Padding";
    
    private Key key;
    private Cipher cipher;

    public static void main(String[] args) {

        String content = "hello world!";

        String key = "d959caadac9b13dcb3e609440135cf54";
        String iv = "1234567812345678";

        // 加密字符串
        System.out.println("加密前的：" + content);
        System.out.println("加密密钥：" + key);

        // 加密方法
        AES aes = new AES();
        String enc = aes.encrypt(content.getBytes(), key.getBytes(), iv.getBytes());
        System.out.println("加密后的内容：" + enc);

        // 解密方法
        String dec = aes.decrypt(enc, key.getBytes(), iv.getBytes());
        System.out.println("解密后的内容：" + dec);
    }

    public void init(byte[] keyBytes) {

        // 如果密钥不足16位，那么就补足.  这个if 中的内容很重要
        int base = 16;
        if (keyBytes.length % base != 0) {
            int groups = keyBytes.length / base + (keyBytes.length % base != 0 ? 1 : 0);
            byte[] temp = new byte[groups * base];
            Arrays.fill(temp, (byte) 0);
            System.arraycopy(keyBytes, 0, temp, 0, keyBytes.length);
            keyBytes = temp;
        }
        // 初始化
        Security.addProvider(new BouncyCastleProvider());
        // 转化成JAVA的密钥格式
        key = new SecretKeySpec(keyBytes, KEY_ALGORITHM);
        try {
            // 初始化cipher
            cipher = Cipher.getInstance(algorithmStr, "BC");
        } catch (NoSuchAlgorithmException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (NoSuchPaddingException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (NoSuchProviderException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }
    /**
     * 加密方法
     *
     * @param content
     *            要加密的字符串
     * @param keyBytes
     *            加密密钥
     * @return
     */
    public String encrypt(byte[] content, byte[] keyBytes, byte[] iv) {
        String encryptedText = null;
        init(keyBytes);
        System.out.println("IV：" + new String(iv));
        try {
            cipher.init(Cipher.ENCRYPT_MODE, key, new IvParameterSpec(iv));
            encryptedText = Base64.getEncoder().encodeToString(cipher.doFinal(content));
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return encryptedText;
    }

    /**
     * 解密方法
     *
     * @param encryptedData
     *            要解密的字符串
     * @param keyBytes
     *            解密密钥
     * @return
     */
    public String decrypt(String encryptedData, byte[] keyBytes, byte[] iv) {
        String encryptedText = null;
        init(keyBytes);
        System.out.println("IV：" + new String(iv));
        try {
            cipher.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(iv));
            encryptedText = new String(cipher.doFinal(Base64.getDecoder().decode(encryptedData)));
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return encryptedText;
    }
}

File: aes/AES.go
package main

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"log"
)

func main() {
	key := []byte("d959caadac9b13dcb3e609440135cf54")
	iv := []byte("1234567812345678")

	content := "hello world!"

	c, _ := Encrypt([]byte(content), key, iv)
	b, _ := Decrypt(c, key, iv)
	p := string(b)

	log.Println("加密后的内容：", c)
	log.Println("解密后的内容：", p)
}

func Encrypt(plainText, key, iv []byte) (string, error) {
	block, err := aes.NewCipher(key)
	//log.Println("blockSize == ", block.BlockSize())
	if err != nil {
		return "", err
	}
	plainText = PKCS7Padding(plainText, block.BlockSize())
	cipherText := make([]byte, len(plainText))
	cipher.NewCBCEncrypter(block, iv).CryptBlocks(cipherText, plainText)
	return base64.StdEncoding.EncodeToString(cipherText), nil
}

func Decrypt(cipherText string, key, iv []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	cipherBytes, _ := base64.StdEncoding.DecodeString(cipherText)
	plainText := make([]byte, len(cipherBytes))
	cipher.NewCBCDecrypter(block, iv).CryptBlocks(plainText, cipherBytes)
	plainText = PKCS7Unpadding(plainText, block.BlockSize())
	return plainText, nil
}

func PKCS7Padding(text []byte, blockSize int) []byte {
	padding := blockSize - len(text)%blockSize
	padText := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(text, padText...)
}

func PKCS7Unpadding(text []byte, blockSize int) []byte {
	length := len(text)
	unpadding := int(text[length-1])
	return text[:(length - unpadding)]
}
